src/audio/dai-zephyr.c

```diff
@@ -1280,6 +1346,128 @@ static void dai_report_xrun(struct dai_data *dd, struct comp_dev *dev, uint32_t
 	buffer_release(buf_c);
 }
 
+/* process and copy stream data from multiple DMA source buffers to sink buffer */
+int dai_zephyr_multi_endpoint_copy(struct dai_data **dd, struct comp_dev *dev,
+				   struct comp_buffer *multi_endpoint_buffer,
+				   int num_endpoints)
+{
+	struct comp_buffer __sparse_cache *buf_c;
+	uint32_t avail_bytes = UINT32_MAX;
+	uint32_t free_bytes = UINT32_MAX;
+	uint32_t frames;
+	uint32_t src_frames, sink_frames;
+	uint32_t frame_bytes;
+	int ret, i;
+	int direction;
+
+	if (!num_endpoints || !dd || !multi_endpoint_buffer)
+		return 0;
+
+	buf_c = buffer_acquire(dd[0]->dma_buffer);
+	frame_bytes = audio_stream_frame_bytes(&buf_c->stream);
+	buffer_release(buf_c);
+
+	direction = dev->direction;
+
+	/* calculate min available/free from all endpoint DMA buffers */
+	for (i = 0; i < num_endpoints; i++) {
+		struct dma_status stat;
+
+		/* get data sizes from DMA */
+		ret = dma_get_status(dd[i]->chan->dma->z_dev, dd[i]->chan->index, &stat);
+		switch (ret) {
+		case 0:
+			break;
+		case -EPIPE:
+			/* DMA status can return -EPIPE and current status content if xrun occurs */
+			if (direction == SOF_IPC_STREAM_PLAYBACK)
+				comp_dbg(dev, "dai_zephyr_multi_endpoint_copy(): dma_get_status() underrun occurred, endpoint: %d ret = %u",
+					 i, ret);
+			else
+				comp_dbg(dev, "dai_zephyr_multi_endpoint_copy(): dma_get_status() overrun occurred, enpdoint: %d ret = %u",
+					 i, ret);
+			break;
```

Comment:
If we break on L1389, free_bytes and avail_bytes can in theory be INT32_MAX when we exit the loop?

Reply:
@ranj063 But we are not doing 
````
	avail_bytes = MIN(avail_bytes, stat.pending_length);
		free_bytes = MIN(free_bytes, stat.free); 
``` 
... if we break. Shouldn't these be modified in case of xrun?